smarty/sampler_smirky.py

```diff
@@ -500,7 +502,8 @@ def best_match_reference_types(self, typelist):
 
         # Get current types and reference types
         current_typenames = [ typename for (smirks, typename) in typelist ]
-        reference_typenames = [ typename for (smirks, typename) in self.reference_types ]
+        reference_typenames = [ typename for typename, smirks in \
+                self.reference_typename_dict.items()]
```

Comment:
These changes were less obvious, but I thought it was redundant to have `self.reference_types` and `self.reference_typename_dict` since they stored the same information.